enhancements/security/openshift-image-policy.md

```diff
@@ -0,0 +1,359 @@
+```
+
+#### Installer
+
+As discussed in [the *Proposal* section](#proposal), the installer will grow a new, feature-gated `internalTestingImagePolicy` property driving ClusterVersion `overrides` extension.
+When set to `Disable`, the following `overrides` entry will be added to the ClusterVersion manifest:
+
+```yaml
+kind: ClusterImagePolicy
+group: config.openshift.io
+name: openshift
+unmanaged: true
+```
+
+##### Installer callers
+
+Hive, the assisted installer, and other actors who create install-configs will need to learn to set `internalTestingImagePolicy: Disabled` if they wish to test unsigned releases that enable the feature-gate.
+
+#### Implementation order
+
+To reduce risk, the recommended implementation order is:
+
+1. Machine-config operator [lifts the release payload scope restriction](#api-extensions).
+2. [Parallel Cincinnati work][OTA-1267] allows ART to push Sigstore signatures to `quay.io/openshift-release-dev/ocp-release:sha256-....sig`.
+3. Testing [the planned ClusterImagePolicy](#machine-config-operator) confirms that the approach works, possibly including [restricted-network use](#restricted-network-use).
+    If testing turns up gaps, round to see if we want to pivot or abandon this effort.
+4. [Installer `internalTestingImagePolicy` property](#installer).
+5. Tooling interacting with unsigned `quay.io/openshift-release-dev/ocp-release` release images ([CI, QE, assisted-installer, Hive, others?](#installer-callers)) updated to set `internalTestingImagePolicy`.
+6. Machine-config operator [adds the `openshift` ClusterImagePolicy manifest](#machine-config-operator).
+
+### Risks and Mitigations
+
+#### Changing our minds
+
+Not much risk for connected clusters.
+If we decide that the policy is not helpful, we can remove from future releases via a [deletion annotation](/enhancements/update/object-removal-manifest-annotation.md).
+
+#### Signing key compromise
+
+4.odd releases are currently supported [for 18 months after GA][maintenance-support].
+There's a risk with a single trusted key that the key 4.17.0 trusts is compromised before 4.17 goes end-of-life.
+With this single-key proposal, the only way to update those clusters if a key is revoked would be the process for disabling the image policy is described in [the *Upgrade / Downgrade Strategy* section](#upgrade--downgrade-strategy), and have cluster administrators install a replacement policy with the new key, after somehow establishing trust in that new key.
```

Comment:
> Cincinnati could direct clusters through the version which updates the policy, so that the following versions would only need to be signed with the new key.

The issue with this approach is that 4.17.middle will certainly have bugs, potentially serious bugs, which are fixed in 4.17.tip.  And we occasionally have regressions like [`CephCapDropPanic`][1] which impact updates within a z-stream.  We don't want 4.17.0 to have to choose between updating into a regression or breaking the Sigstore chain.  Or to delay fixing the other bugs fixed in 4.17.tip.  So we need to keep signing 4.17.tip with a system that 4.17.0 trusts out of the box, or provide some kind of hotfix image policy that can get that 4.17.0 straight through to 4.17.tip.

> (One possible problem with this approach is that OpenShift Image Streams, and the integrated registry, only record 6 simple signing signatures per image, and we use about 6 for many RHEL images (2 RH registries × tags (latest, 9, 9.5.1–.…) )

`latest`, `9`, `9.5.1-...` are all separate tags right?  Maybe they're all pointing at the same image?  `ocp-release` has many, many separate release images (one for each CPU architecture for 4.y.z), see [here][2] for page after page of one-per-digest OpenPGP signatures for various `quay.io/openshift-release-dev/ocp-release` images, possibly up near 10k by now.

[1]: https://issues.redhat.com/browse/COS-2781
[2]: https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/

Reply:
> So we need to keep signing 4.17.tip with a system that 4.17.0 trusts out of the box

That seems fine to me — I suppose it’s possible we could somehow _lose_ the key, but that’s an entirely different scenario, and in that case the cluster _will_ need a manual configuration change no matter what.

Reply:
> `latest`, `9`, `9.5.1-...` are all separate tags right? Maybe they're all pointing at the same image?

Yes: we tag the same image into several tags, and sign all of them so, that users can run (now real names:`registry.access.redhat.com/ubi9/ubi:latest`, `registry.access.redhat.com/ubi9/ubi:9.4
`, or `registry.access.redhat.com/ubi9/ubi:9.4-947.1716476138`).

The limitation is 6 signatures per “deduplicated image”; not per repo.

---

>  see [here](https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/) for page after page of one-per-digest OpenPGP signatures

A random spot-check suggests that there is typically only one signature per image (= per image digest = per directory in that URL); if that’s consistently the case, we are fine for that image.

Reply:
> Yes: we tag the same image into several tags, and sign all of them...

Release images should always be by-digest, see openshift/oc#390 for client-side guards.  And the web-console only allows `payload` from the OpenShift Update Service, which we configure to only recommend by-digest pullspecs.  And release-image references to other payload images are also by-digest:

```console
$ oc adm release info --pullspecs quay.io/openshift-release-dev/ocp-release@sha256:d93db95ec59b2f51ff45fdecd54b4dd9144521e3d099fef151c189598200cf24 
Name:           4.16.0-rc.2
Digest:         sha256:d93db95ec59b2f51ff45fdecd54b4dd9144521e3d099fef151c189598200cf24
Created:        2024-05-17T13:49:31Z
...
Images:
  NAME                                           PULL SPEC
  agent-installer-api-server                     quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:fda26b66691690687c3cac90a01eb238f0671a763f73e0ec0a9d23672f938757
  agent-installer-csr-approver                   quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:f46665f0efe030dc86cad8608bf657c83d0687dc45cf84cbf2f7f9e489859f1b
  agent-installer-node-agent                     quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:01072d5f00788246db2585e4410376b664d864077bcda99faeb73e84d6876a53
...
```

So yeah, should be all good there.

I've pushed d7dc3f54 -> 22fb271b, adding a paragraph about continued 4.17.tip signing, a pointer to containers/image#2319, and a new *Signing key destruction* section covering the lost-key risk you just brought up.  I'll mark this thread resolved on that, but please open a new thread if you think  I still need further changes.